File: test_dir/bubble_sort.py
def bubble_sort(arr):
    """
    冒泡排序算法
    时间复杂度: O(n²)
    空间复杂度: O(1)
    稳定排序
    """
    n = len(arr)
    for i in range(n):
        # 提前退出标志：如果某一轮没有发生交换，说明已经有序
        swapped = False
        for j in range(n - 1 - i):
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                swapped = True
        if not swapped:
            break
    return arr


if __name__ == "__main__":
    # 测试用例
    test_cases = [
        [64, 34, 25, 12, 22, 11, 90],
        [5, 1, 4, 2, 8],
        [3, 3, 3],
        [1],
        [],
        [10, 9, 8, 7, 6, 5, 4, 3, 2, 1],
    ]

    for case in test_cases:
        print(f"排序前: {case}")
        print(f"排序后: {bubble_sort(case)}")
        print("-" * 30)

File: bubble_sort.c
#include <stdio.h>

// 冒泡排序
void bubbleSort(int arr[], int n) {
    for (int i = 0; i < n - 1; i++) {
        // 标志位：若本轮未发生交换，说明已有序，提前结束
        int swapped = 0;
        for (int j = 0; j < n - 1 - i; j++) {
            if (arr[j] > arr[j + 1]) {
                int tmp = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = tmp;
                swapped = 1;
            }
        }
        if (!swapped) {
            break;
        }
    }
}

// 打印数组
void printArray(int arr[], int n) {
    for (int i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");
}

int main() {
    int arr[] = {64, 34, 25, 12, 22, 11, 90};
    int n = sizeof(arr) / sizeof(arr[0]);

    printf("排序前: ");
    printArray(arr, n);

    bubbleSort(arr, n);

    printf("排序后: ");
    printArray(arr, n);

    return 0;
}
